crates/red_knot_python_semantic/src/module_resolver/resolver.rs

```diff
@@ -569,24 +569,19 @@ fn resolve_name(db: &dyn Db, name: &ModuleName) -> Option<(SearchPath, File, Mod
 
                 package_path.push(module_name);
 
-                // Must be a `__init__.pyi` or `__init__.py` or it isn't a package.
-                let kind = if package_path.is_directory(&resolver_state) {
-                    package_path.push("__init__");
-                    ModuleKind::Package
-                } else {
-                    ModuleKind::Module
-                };
-
-                // TODO Implement full https://peps.python.org/pep-0561/#type-checker-module-resolution-order resolution
-                if let Some(stub) = package_path.with_pyi_extension().to_file(&resolver_state) {
-                    return Some((search_path.clone(), stub, kind));
+                // Try to resolve a stub/module with the current name
+                let file_module = resolve_file_module(&package_path, &resolver_state);
+
+                // Then check if a regular package exists with the same name (it takes
+                // precedence)
+                package_path.push("__init__");
+                if let Some(regular_package) = resolve_file_module(&package_path, &resolver_state) {
+                    return Some((search_path.clone(), regular_package, ModuleKind::Package));
                 }
 
-                if let Some(module) = package_path
-                    .with_py_extension()
-                    .and_then(|path| path.to_file(&resolver_state))
-                {
-                    return Some((search_path.clone(), module, kind));
+                // If we found a module (but no package), return it
+                if let Some(module) = file_module {
+                    return Some((search_path.clone(), module, ModuleKind::Module));
```

Comment:
This logic can be simplified slightly if we make the `ModulePath::join()` function in `path.rs` public (it already exists, but it's currently a test-only private function). Something like this should work (the diff is relative to your PR branch):

```diff
diff --git a/crates/red_knot_python_semantic/src/module_resolver/path.rs b/crates/red_knot_python_semantic/src/module_resolver/path.rs
index a49d1dc20..76c00b1c2 100644
--- a/crates/red_knot_python_semantic/src/module_resolver/path.rs
+++ b/crates/red_knot_python_semantic/src/module_resolver/path.rs
@@ -59,6 +59,13 @@ impl ModulePath {
         self.relative_path.push(component);
     }
 
+    #[must_use]
+    pub(crate) fn join(&self, component: &str) -> ModulePath {
+        let mut result = self.clone();
+        result.push(component);
+        result
+    }
+
     #[must_use]
     pub(super) fn is_directory(&self, resolver: &ResolverContext) -> bool {
         let ModulePath {
@@ -634,15 +641,6 @@ mod tests {
 
     use super::*;
 
-    impl ModulePath {
-        #[must_use]
-        fn join(&self, component: &str) -> ModulePath {
-            let mut result = self.clone();
-            result.push(component);
-            result
-        }
-    }
-
     impl SearchPath {
         fn join(&self, component: &str) -> ModulePath {
             self.to_module_path().join(component)
diff --git a/crates/red_knot_python_semantic/src/module_resolver/resolver.rs b/crates/red_knot_python_semantic/src/module_resolver/resolver.rs
index 50aaf9dbd..294ad477d 100644
--- a/crates/red_knot_python_semantic/src/module_resolver/resolver.rs
+++ b/crates/red_knot_python_semantic/src/module_resolver/resolver.rs
@@ -569,18 +569,13 @@ fn resolve_name(db: &dyn Db, name: &ModuleName) -> Option<(SearchPath, File, Mod
 
                 package_path.push(module_name);
 
-                // Try to resolve a stub/module with the current name
-                let file_module = resolve_file_module(&package_path, &resolver_state);
-
-                // Then check if a regular package exists with the same name (it takes
-                // precedence)
-                package_path.push("__init__");
-                if let Some(regular_package) = resolve_file_module(&package_path, &resolver_state) {
+                // Regular packages take first priority:
+                if let Some(regular_package) = resolve_file_module(&package_path.join("__init__"), &resolver_state) {
                     return Some((search_path.clone(), regular_package, ModuleKind::Package));
                 }
 
-                // If we found a module (but no package), return it
-                if let Some(module) = file_module {
+                // If there's a single-file module (but no package), return it
+                if let Some(module) = resolve_file_module(&package_path, &resolver_state) {
                     return Some((search_path.clone(), module, ModuleKind::Module));
                 }
```

Reply:
The reason for using `push` is because it avoids allocating a new path.

Reply:
My main point is that if the path resolves to a regular package, we're doing an unnecessary `resolve_file_module()` call. I doubt it matters much in terms of performance but I think it could be written slightly more clearly. If we want to avoid allocating, we could add a `.pop()` method:

```diff
--- a/crates/red_knot_python_semantic/src/module_resolver/path.rs
+++ b/crates/red_knot_python_semantic/src/module_resolver/path.rs
@@ -59,6 +59,10 @@ impl ModulePath {
         self.relative_path.push(component);
     }
 
+    pub(crate) fn pop(&mut self) {
+        self.relative_path.pop();
+    }
+
     #[must_use]
     pub(super) fn is_directory(&self, resolver: &ResolverContext) -> bool {
         let ModulePath {
diff --git a/crates/red_knot_python_semantic/src/module_resolver/resolver.rs b/crates/red_knot_python_semantic/src/module_resolver/resolver.rs
index 68f08118e..c563d0b80 100644
--- a/crates/red_knot_python_semantic/src/module_resolver/resolver.rs
+++ b/crates/red_knot_python_semantic/src/module_resolver/resolver.rs
@@ -569,9 +569,6 @@ fn resolve_name(db: &dyn Db, name: &ModuleName) -> Option<(SearchPath, File, Mod
 
                 package_path.push(module_name);
 
-                // Try to resolve a stub/module with the current name
-                let file_module = resolve_file_module(&package_path, &resolver_state);
-
                 // Then check if a regular package exists with the same name (it takes
                 // precedence)
                 package_path.push("__init__");
@@ -579,8 +576,9 @@ fn resolve_name(db: &dyn Db, name: &ModuleName) -> Option<(SearchPath, File, Mod
                     return Some((search_path.clone(), regular_package, ModuleKind::Package));
                 }
 
-                // If we found a module (but no package), return it
-                if let Some(module) = file_module {
+                // If we can find a module (but no package), return it
+                package_path.pop();
+                if let Some(module) = resolve_file_module(&package_path, &resolver_state) {
                     return Some((search_path.clone(), module, ModuleKind::Module));
                 }
```

Reply:
I think I find either options (`join` or `pop`) clearer than the existing implementation